### PR TITLE
Implement haltWhen

### PR DESCRIPTION
**Pull request by luis3m**
Part of #4886

### CODE REVIEW

**Comment by luis3m** [comment]
I'm going to rewrite this one without `.toPull`

**Comment by luis3m** [comment]
Done

**Comment by luis3m** [comment]
Actually, I can simplify. Changing.

**Comment by luis3m** [comment]
Done, @iravid

File: streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala [hunk 1]
```diff
@@ -1238,8 +1238,30 @@ class ZStream[-R, +E, +A](val channel: ZChannel[R, Any, Any, Any, E, Chunk[A], A
...
+            )
+
+          case Some(exit) =>
+            exit.fold(ZChannel.halt(_), _ => ZChannel.unit)
+        }
+      }
+
+    new ZStream(
+      ZChannel.unwrap {
+        io.fork.map { fiber =>
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala:1259]: We should use forkManaged here

Comment (luis3m) [reply streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala]: Done, @iravid

**Review by iravid — APPROVED** [review]

**Pull request merged**